libsolidity/ast/ASTUtils.h

```diff
@@ -33,4 +35,7 @@ VariableDeclaration const* rootConstVariableDeclaration(VariableDeclaration cons
 /// Returns true if the constant variable declaration is recursive.
 bool isConstantVariableRecursive(VariableDeclaration const& _varDecl);
 
+/// Locates closest AST node match within a SourceUnit given its absolute file offset or nullptr if none matched.
+ASTNode const* locateASTNode(int _pos, SourceUnit const& _sourceUnit);
```

Comment:
It should mention that if there are multiple matches it just takes the one it comes across first.

Reply:
But actually, maybe it would be better to have it look at all matches and choose the shortest one? For example I think you do not want to get the whole function definition as a result when you're only pointing at a local variable.

Reply:
If the position is pointing at a variable then you won't get the whole function AST node back but the `Identifier` AST node of that variable. Mind, our AST currently is not **that** precise. There is room for improvement, such as we don't store all identifiers as AST node `Identifier` but instead just as `std::string` - that is why some while ago I introduced `nameLocation()` along with `location()` to get access to the name's location. I'd like to have that changed to a real AST node at some point.

> But actually, maybe it would be better to have it look at all matches and choose the shortest one? 

I think that depends on the use-case. If one needs more fine grained control over finding AST nodes, maybe there should be a specialized finder then. Currently this API suffices the current use I think.

Reply:
> If the position is pointing at a variable then you won't get the whole function AST node back but the Identifier AST node of that variable.

But how does it do that? Is it because of visiting order? If so, it's not obvious at all.

Reply:
Answered in https://github.com/ethereum/solidity/pull/12304#discussion_r755250156.